### PR TITLE
Feature questions

### PR DESCRIPTION
**Pull request by techaffinity**
Polls feature - Admin can create questions, choices and map questions to the pages. Polls associated to the pages will get displayed in the front end. Basic UI alignment is covered in the polls display. Conflicts resolved and adapt to new folder structure. Polls have been created as a separate app and included in the settings. We have checked for the flake8 standards for this poll feature.

### CODE REVIEW

**Comment by techaffinity** [comment]
Simon, we will check for the standards in another revision and update the same here.

**Comment by smn** [comment]
I merged in #4 to remove some of the commits that were already reviewed earlier.

Going forward please use `git flow` to manage your branches, it makes it easier to review & keep track of work in progress. It's documented in our [ways of working](http://ways-of-working.readthedocs.org/en/latest/process.html#example-flow) document but please shout if we need to explain this again separately.

File: ndohyep/polls/detail.html [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+	<form action="{% url 'vote' question.id %}" method="post">
+		{% csrf_token %}
+		{% for choice in question.choice_set.all %}
+		    <input type="radio" name="choice" id="choice{{ forloop.counter }}" value="{{ choice.id }}" />
+		    <label for="choice{{ forloop.counter }}">{{ choice.choice_text }}</label><br />
+		{% endfor %}
+		<input type="submit" value="Vote" />
+	</form>
+	<a href="{% url 'results' question.id %}">See Results</a>
+{% endblock %}
```
Comment (smn) [inline ndohyep/polls/detail.html]: missing new line.

File: ndohyep/polls/index.html [hunk 2]
```diff
@@ -0,0 +1,10 @@
+{% load i18n %}
+{% if latest_question_list %}
+    <ul>
+    {% for question in latest_question_list %}
+        <li><a href="/polls/{{ question.id }}/">{{ question.question_text }}</a></li>
+    {% endfor %}
+    </ul>
+{% else %}
+    <p>No polls are available.</p>
+{% endif %}
```
Comment (smn) [inline ndohyep/polls/index.html]: Missing new line.

File: ndohyep/polls/index.html [hunk 3]
```diff
@@ -0,0 +1,10 @@
+{% load i18n %}
+{% if latest_question_list %}
+    <ul>
+    {% for question in latest_question_list %}
+        <li><a href="/polls/{{ question.id }}/">{{ question.question_text }}</a></li>
+    {% endfor %}
+    </ul>
+{% else %}
+    <p>No polls are available.</p>
```
Comment (smn) [inline ndohyep/polls/index.html]: `{% trans "..." %}`

File: ndohyep/polls/poll_page.html [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+		<div class="clear-block">
+		<h1>{{ question.question_text }}</h1>
+			{% if error_message %}<p><strong>{{ error_message }}</strong></p>{% endif %}
+				<form action="{% url 'vote' question.id %}" method="post">
+					{% csrf_token %}
+					{% for choice in question.choice_set.all %}
+					    <input type="radio" name="choice" id="choice{{ forloop.counter }}" value="{{ choice.id }}" />
+					    <label for="choice{{ forloop.counter }}">{{ choice.choice_text }}</label><br />
+					{% endfor %}
+					<input class="m0" type="submit" value="Vote" />
```
Comment (smn) [inline ndohyep/polls/poll_page.html]: `{% trans "..." %}`

File: ndohyep/polls/poll_page.html [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+				</form>
+				<a class="link" href="{% url 'results' question.id %}">See Results</a>
```
Comment (smn) [inline ndohyep/polls/poll_page.html]: `{% trans "..." %}`

File: ndohyep/polls/poll_page.html [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+				<div class="clear-block">
+		<h1>{{ question.question_text }}</h1>
+			{% if error_message %}<p><strong>{{ error_message }}</strong></p>{% endif %}
+				<form action="{% url 'vote' question.id %}" method="post">
+					{% csrf_token %}
+					{% for choice in question.choice_set.all %}
+					    <input type="radio" name="choice" id="choice{{ forloop.counter }}" value="{{ choice.id }}" />
+					    <label for="choice{{ forloop.counter }}">{{ choice.choice_text }}</label><br />
+					{% endfor %}
+					<input class="m0" type="submit" value="Vote" />
```
Comment (smn) [inline ndohyep/polls/poll_page.html]: `{% trans "..." %}`

File: ndohyep/polls/poll_page.html [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+				</form>
+				<a class="link" href="{% url 'results' question.id %}">See Results</a>
```
Comment (smn) [inline ndohyep/polls/poll_page.html]: `{% trans "..." %}`

File: ndohyep/polls/poll_widget.html [hunk 8]
```diff
@@ -0,0 +1,11 @@
+{% load i18n %}
+
+{% if latest_question_list %}
+    <ul>
+    {% for question in latest_question_list %}
+        <li><a href="/polls/{{ question.id }}/">{{ question.question_text }}</a></li>
+    {% endfor %}
+    </ul>
+{% else %}
+    <p>No polls are available.</p>
```
Comment (smn) [inline ndohyep/polls/poll_widget.html]: `{% trans "..." %}`

File: ndohyep/polls/results.html [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+<div class="clear-block">
+<h1>{{ question.question_text }}</h1>
+
+<ul>
+{% for choice in question.choice_set.all %}
+    <li>{{ choice.choice_text }} -- {{ choice.votes }} vote{{ choice.votes|pluralize }}</li>
+{% endfor %}
+</ul>
+
+<a href="{% url 'detail' question.id %}">Vote again?</a>
```
Comment (smn) [inline ndohyep/polls/results.html]: `{% trans "..." %}`

File: ndohyep/polls/results.html [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+
+{% load static core_tags %}
+
+{% block content %}
+<div class="clear-block">
+<h1>{{ question.question_text }}</h1>
+
+<ul>
+{% for choice in question.choice_set.all %}
+    <li>{{ choice.choice_text }} -- {{ choice.votes }} vote{{ choice.votes|pluralize }}</li>
```
Comment (smn) [inline ndohyep/polls/results.html]: `{% trans "..." %}`

File: polls/models.py [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+class LanguagePage(LanguagePage):
+    LanguagePage.subpage_types = [
+        'core.HomePage',
+        'core.SectionPage',
+        'core.FooterPage',
+        'polls.Question'
+    ]
+
+
+class Question(Page):
```
Comment (smn) [inline polls/models.py:19]: @miltontony could you check the approach being taken here around subclassing a `Page` for a `Question` and a `Choice` further down?

File: polls/templatetags/poll_votings.py [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+
+register = template.Library()
+
+
+@register.inclusion_tag('polls/poll_page.html',
+                        takes_context=True)
+def poll_page(context, pk=None, page=None):
+    context = copy(context)
+    polls = Question.objects.filter(page=page.id)
+    if polls:
```
Comment (smn) [inline polls/templatetags/poll_votings.py]: `polls.exists()`

File: polls/templatetags/poll_votings.py [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+        context.update({
+            'questions': polls
+        })
+    else:
+        context.update({
+            'error': True,
+        })
```
Comment (smn) [inline polls/templatetags/poll_votings.py]: Why do we need to return an `error` here? Why not an empty `[]` for the `questions` key? If a `Page` doesn't have any set questions then we can handle that in the template without needing a separate context variable set, I think that only makes the template more complex than necessary.

File: polls/urls.py [hunk 14]
```diff
@@ -0,0 +1,13 @@
...
+
+urlpatterns = [
+    url(r'^$', views.IndexView.as_view(), name='index'),
+    url(r'^(?P<pk>[0-9]+)/$', views.DetailView.as_view(), name='detail'),
+    url(
+        r'^(?P<pk>[0-9]+)/results/$',
+        views.ResultsView.as_view(),
+        name='results'
+    ),
+    url(r'^(?P<question_id>[0-9]+)/vote/$', views.vote, name='vote'),
```
Comment (smn) [inline polls/urls.py]: It doesn't make a huge difference but sometimes using a `\d` (digit) as a placeholder in regex is easier than `[0-9]`:

``` python
>>>> match = re.match(r'(?P<pk>\d+)/$', '1234/')
>>>> match.groupdict()
{'pk': '1234'}
>>>>
```

**Comment by smn** [comment]
Looping in @bruskiza & @saeedmb8 so they're aware of the review going on and can follow along as needed.

File: polls/views.py [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+    except (KeyError, Choice.DoesNotExist):
+        # Redisplay the question voting form.
+        return render(request, 'polls/detail.html', {
+            'question': question,
+            'error_message': "You didn't select a choice.",
+        })
+    else:
+        selected_choice.votes += 1
+        selected_choice.save()
+        return HttpResponseRedirect(reverse('results', args=(question.id,)))
```
Comment (smn) [inline polls/views.py:41]: Why is this not handled in the `DetailView` like a normal form would in Django? Any specific reason?
That way we can handle the error reporting in that template and also we could avoid the race condition at line 39 (`selected_choice.votes += 1`) by using a `filter()` and an `update()` query using an [`F()` expression](https://docs.djangoproject.com/en/1.7/ref/models/queries/#django.db.models.F).

File: polls/views.py [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+
+def vote(request, question_id):
+    question = get_object_or_404(Question, pk=question_id)
+    try:
+        selected_choice = question.choice_set.get(pk=request.POST['choice'])
+    except (KeyError, Choice.DoesNotExist):
+        # Redisplay the question voting form.
+        return render(request, 'polls/detail.html', {
+            'question': question,
+            'error_message': "You didn't select a choice.",
```
Comment (smn) [inline polls/views.py]: `_()`

File: ndohyep/polls/index.html [hunk 17]
```diff
@@ -0,0 +1,10 @@
+{% load i18n %}
+{% if latest_question_list %}
+    <ul>
+    {% for question in latest_question_list %}
+        <li><a href="/polls/{{ question.id }}/">{{ question.question_text }}</a></li>
```
Comment (smn) [inline ndohyep/polls/index.html]: `{% url ... %}`

File: ndohyep/polls/poll_widget.html [hunk 18]
```diff
@@ -0,0 +1,11 @@
+{% load i18n %}
+
+{% if latest_question_list %}
+    <ul>
+    {% for question in latest_question_list %}
+        <li><a href="/polls/{{ question.id }}/">{{ question.question_text }}</a></li>
```
Comment (smn) [inline ndohyep/polls/poll_widget.html]: `{% url ... %}`

File: ndohyep/polls/poll_page.html [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+				</div>
+		{% endfor %}
+	{% else %}
+		<p>No polls are available.</p>
+
+	{% endif %}
+{% endblock %}
```
Comment (smn) [inline ndohyep/polls/poll_page.html]: missing new line.

File: ndohyep/polls/poll_widget.html [hunk 20]
```diff
@@ -0,0 +1,11 @@
...
+    {% endfor %}
+    </ul>
+{% else %}
+    <p>No polls are available.</p>
+{% endif %}
```
Comment (smn) [inline ndohyep/polls/poll_widget.html]: missing new line.

**Comment by smn** [comment]
@techaffinity thanks for the pull request, I have reviewed it. There were still a few missing `{% trans ... %}` and `{% url ... %}` bits but significantly less than #4. Also thanks for making sure pep8 compliance, there was one problem with `settings/base.py` having an extra new line which I've attended to.

However there does seem to be a problem with your migrations though, check the travis build logs for the details.

File: polls/models.py (hunk 11 above)
Comment (miltontony) [inline polls/models.py:19]: @smn I see the _why_. This whole `Page` model that Wagtail enforces is one of the reasons I'm unhappy with Wagtail. Until we've done more research into ways of extending it, this approach will have to do.

**Comment by bruskiza** [comment]
Sticking a foot in here... is it not a wagtail version thing? I may have similar issues.

**Comment by smn** [comment]
@techaffinity please pull in the latest changes before you continue work on this. @miltontony and myself have been busy fixing a bunch of things in the mean time.

File: polls/templatetags/poll_votings.py (hunk 13 above)
Comment (techaffinity) [inline polls/templatetags/poll_votings.py]: The extra context variable was removed and the questions empty handled in the template level.

**Comment by smn** [comment]
@techaffinity you've broken the build, please fix it. I will review once the build passes. I suspect it's related to @miltontony's [earlier comment](https://github.com/praekelt/molo-ndoh-yep/pull/5#commitcomment-12465976)

**Comment by techaffinity** [comment]
We have checked this build in our end and we could see this build now that have been passed without any migration issue.

File: ndohyep/polls/index.html (hunk 17 above)
Comment (smn) [inline ndohyep/polls/index.html]: :point_up:

File: ndohyep/polls/poll_widget.html [hunk 21]
```diff
@@ -0,0 +1,11 @@
...
+
+{% if latest_question_list %}
+    <ul>
+    {% for question in latest_question_list %}
+        <li><a href="/polls/{{ question.id }}/">{{ question.question_text }}</a></li>
+    {% endfor %}
+    </ul>
+{% else %}
+    <p>{% trans "No polls are available." %}</p>
+{% endif %}
```
Comment (smn) [inline ndohyep/polls/poll_widget.html]: :point_up:

File: ndohyep/static/css/ndohyep.css [hunk 22]
```diff
@@ -401,4 +407,10 @@ a.backtotop {padding:20px; display:block; clear:both; color:white; text-transfor
...
     text-decoration: none; }
 
-                    
+.m0 {
+    margin-right: 0 !important; }
+
+a.link {
+    color: #000;
+    font-size: 14px;
+}
```
Comment (smn) [inline ndohyep/static/css/ndohyep.css:416]: :point_up:

**Comment by smn** [comment]
I'm going to go ahead and merge this in. If there's more work to be done please create a separate pull-request.

**Pull request merged**